### PR TITLE
Add support for client-specified headers in connector_extra_data

### PR DESCRIPTION
**Pull request by danlu-square**

### CODE REVIEW

**Review by shellderp — COMMENTED** [review]
looks good, maybe we can add unit tests?

**Comment by danlu-square** [comment]
> maybe we can add unit tests?

Will do

File: client/src/main/kotlin/app/cash/backfila/client/Connectors.kt [hunk 1]
```diff
@@ -5,6 +5,8 @@ object Connectors {
   const val ENVOY = "ENVOY"
 }
 
-data class HttpConnectorData(val url: String)
+data class HttpHeader(val name: String, val value: String)
 
-data class EnvoyConnectorData(val clusterType: String)
+data class HttpConnectorData(val url: String, val headers: List<HttpHeader> = listOf())
+
+data class EnvoyConnectorData(val clusterType: String, val headers: List<HttpHeader> = ArrayList())
```
Comment (shellderp) [inline client/src/main/kotlin/app/cash/backfila/client/Connectors.kt:12]:
```suggestion
data class EnvoyConnectorData(val clusterType: String, val headers: List<HttpHeader> = listOf())
```

**Comment by danlu-square** [comment]
Added some tests exercising the interceptor and `connector_extra_data` -> `HttpHeader` parsing

File: service/src/main/kotlin/app/cash/backfila/client/interceptors/OkHttpClientSpecifiedHeadersInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package app.cash.backfila.client.interceptors
+
+import app.cash.backfila.client.HttpHeader
+import okhttp3.Interceptor
+import okhttp3.Response
+
+internal class OkHttpClientSpecifiedHeadersInterceptor internal constructor(
+  private val headers: List<HttpHeader>?,
```
Comment (shellderp) [inline service/src/main/kotlin/app/cash/backfila/client/interceptors/OkHttpClientSpecifiedHeadersInterceptor.kt:8]: how come this is nullable? if it's null, we shouldn't create the interceptor

Comment (danlu-square) [reply service/src/main/kotlin/app/cash/backfila/client/interceptors/OkHttpClientSpecifiedHeadersInterceptor.kt]: Fair enough, I've switched it to be non-nullable. The calling code already behaved that way.

**Review by shellderp — APPROVED** [review]

File: service/src/main/kotlin/app/cash/backfila/client/EnvoyClientServiceClientProvider.kt [hunk 3]
```diff
@@ -23,23 +24,43 @@ class EnvoyClientServiceClientProvider @Inject constructor(
   lateinit var envoyClientEndpointProvider: EnvoyClientEndpointProvider
 
   override fun validateExtraData(connectorExtraData: String?) {
-    connectorExtraData?.let { adapter().fromJson(connectorExtraData)?.clusterType }
+    connectorExtraData?.let {
+      val fromJson = adapter().fromJson(connectorExtraData)
+      checkNotNull(fromJson, { "Failed to parse HTTP connector extra data JSON" })
+
+      if (!fromJson.headers.isNullOrEmpty()) {
+        for (header in fromJson.headers) {
```
[the remark this answers is not in the document]
Comment (mpawliszyn) [reply service/src/main/kotlin/app/cash/backfila/client/EnvoyClientServiceClientProvider.kt:35]: Yeah I was thinking of a byte size limit. A count of 10 might not be enough in certain cases and I care less about the count and more about the byte size.

[the remark this answers is not in the document]
Comment (mpawliszyn) [reply service/src/main/kotlin/app/cash/backfila/client/EnvoyClientServiceClientProvider.kt:35]: 2k should be plenty?
https://stackoverflow.com/questions/686217/maximum-on-http-header-values

**Review by mpawliszyn — APPROVED** [review]
I'd rather a size than a count limit but looking great otherwise!

File: service/src/main/kotlin/app/cash/backfila/client/EnvoyClientServiceClientProvider.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (danlu-square) [reply service/src/main/kotlin/app/cash/backfila/client/EnvoyClientServiceClientProvider.kt:35]: Cool, done
Tests now include a check on multiple smaller headers as well as a single very large header

File: service/src/main/kotlin/app/cash/backfila/client/BackfilaClientServiceClientProvider.kt [hunk 4]
```diff
@@ -24,6 +24,11 @@ class BackfilaDefaultEndpointConfigModule() : KAbstractModule() {
 }
 
 interface BackfilaClientServiceClientProvider {
+  fun headersSize(headers: List<HttpHeader>): Int {
+    return headers.joinToString("\r\n") { header -> "${header.name}: ${header.value}" }
+      .toByteArray()
+      .size
```
Comment (mpawliszyn) [inline service/src/main/kotlin/app/cash/backfila/client/BackfilaClientServiceClientProvider.kt:30]: I'm also not sure this belongs here? it is more a property of the header than the provider.

File: service/src/main/kotlin/app/cash/backfila/client/interceptors/OkHttpClientSpecifiedHeadersInterceptor.kt [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+    val requestBuilder = chain.request().newBuilder()
+    for (header in headers) {
+      requestBuilder.header(header.name, header.value)
+    }
+    return chain.proceed(requestBuilder.build())
+  }
+
+  companion object {
+    const val MAX_ALLOWED_HEADERS_SIZE = 2048
```
Comment (mpawliszyn) [inline service/src/main/kotlin/app/cash/backfila/client/interceptors/OkHttpClientSpecifiedHeadersInterceptor.kt:24]: Maybe the size function should be a companion of this?

Comment (danlu-square) [reply service/src/main/kotlin/app/cash/backfila/client/interceptors/OkHttpClientSpecifiedHeadersInterceptor.kt:24]: Done

File: service/src/main/kotlin/app/cash/backfila/client/interceptors/OkHttpClientSpecifiedHeadersInterceptor.kt [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+    fun headersSizeWithinLimit(headers: List<HttpHeader>): Boolean {
```
Comment (mpawliszyn) [inline service/src/main/kotlin/app/cash/backfila/client/interceptors/OkHttpClientSpecifiedHeadersInterceptor.kt:26]: Even better!

**Review by mpawliszyn — APPROVED** [review]

**Pull request merged**